Upgrade snap base to core24

farshidtz opened the pull request:
Related issue:
- Closes https://github.com/canonical/docker-snap/issues/204

Related PRs:
- #203 (main implementation)
- #239 (tests)
- #274 
- #290
- #291
- #288
- #289
- #293
- #302
- #303 (docs)




Snap revisions built from the core24 branch are available on the `latest/edge/core24` channel, for amd64 only. To install:
```bash
sudo snap install docker --channel=edge/core24
```

jocado committed:
Nvidia support core24 (#203)

* Bump libnvidia-container and nvidia-container-toolkit to version 1.17.3

* Add initial version of support for uc24, sill alongside uc22 support

* Update comment with correct syntax regarding spec-dirs

* Switch base to core24, and remove core22 based nvidia support

* Remove provider for gpu-2404 to prevent pulling in snap as a dedendency by default

* Replace gpu-2404-wrapper with forked optional version, and disable nvidia-container-toolkit on systems which don't have any detected nvidia support

* Add core22 based nvidia support again, to support both UC24 and UC22

* fix: make sure the gpu wrapper only applies to the required ubuntu versions

* Move nvidia support detection to separate command-chain, and add some basic logging on detected status

* Move some setup logic to post-refresh only, to prevent it being run twice for no good reason

farshidtz committed:
Merge remote-tracking branch 'origin/main' into core24

farshidtz commented:
Testing with 28.0.4 engine: https://github.com/canonical/docker-snap/actions/runs/14194910223

farshidtz committed:
Merge remote-tracking branch 'origin/main' into core24

farshidtz committed:
Activate Nvidia core24/UC24 testing
[skip ci]

nateblack commented:
@farshidtz , can we please have the `latest/edge/core24` updated again? The previous snap revision 3070 that was working (https://github.com/canonical/docker-snap/pull/203) appears to no longer be available. Attempts to use `/latest/edge/core24` just results in using `/latest/edge`. The current snap revision 3192 on `latest/edge` does not work.

`2025-04-14T19:58:25Z systemd[1]: Starting snap.docker.nvidia-container-toolkit.service - Service for snap application docker.nvidia-container-toolkit...
2025-04-14T19:58:25Z docker.nvidia-container-toolkit[3770]: Running on Ubuntu Core system
2025-04-14T19:58:25Z docker.nvidia-container-toolkit[3770]: NVIDIA hardware detected: 01:00.0 VGA compatible controller: NVIDIA Corporation Device 2803 (rev a1)
2025-04-14T19:58:25Z docker.nvidia-container-toolkit[3770]: 01:00.1 Audio device: NVIDIA Corporation Device 22bd (rev a1)
2025-04-14T19:58:25Z docker.nvidia-container-toolkit[3770]: Waiting for device to become available: /dev/nvidiactl
2025-04-14T19:58:25Z docker.nvidia-container-toolkit[3770]: Checking device: 0/10
2025-04-14T19:58:25Z docker.nvidia-container-toolkit[3770]: Device found
2025-04-14T19:58:25Z docker.nvidia-container-toolkit[3770]: NVIDIA ready
2025-04-14T19:58:26Z docker.nvidia-container-toolkit[3899]: time="2025-04-14T19:58:26Z" level=info msg="Auto-detected mode as \"nvml\""
2025-04-14T19:58:26Z docker.nvidia-container-toolkit[3899]: time="2025-04-14T19:58:26Z" level=error msg="failed to generate CDI spec: failed to create device CDI specs: failed to initialize NVML: ERROR_LIBRARY_NOT_FOUND"
2025-04-14T19:58:26Z docker.nvidia-container-toolkit[3770]: WARNING: Conainter Toolkit setup seemed to fail with an error
2025-04-14T19:58:26Z systemd[1]: snap.docker.nvidia-container-toolkit.service: Deactivated successfully.
2025-04-14T19:58:26Z systemd[1]: Finished snap.docker.nvidia-container-toolkit.service - Service for snap application docker.nvidia-container-toolkit.`

farshidtz commented:
@nateblack I've just released a build from this branch. The revision number is 3200.

nateblack commented:
> @nateblack I've just released a build from this branch. The revision number is 3200.

Thank you!

farshidtz commented:
@nateblack apologies for the late response. 

The nvidia logic in this branch is similar to #203, except that the Docker parts have been upgraded. 

The `docker:graphics-core22` plug should not be auto connected, unless you have the `nvidia-core22` installed, which provides the `graphics-core22` slot. Since `docker` and `nvidia-core22` share the same publisher, the interface gets auto connected if they are both installed. Docker snap has no other assertions for auto connecting its graphics plugs.

You can check the interface connections with: `snap connections docker`

nateblack commented:
> @nateblack apologies for the late response.
> 
> The nvidia logic in this branch is similar to #203, except that the Docker parts have been upgraded.
> 
> The `docker:graphics-core22` plug should not be auto connected, **unless you have the `nvidia-core22` installed**, which provides the `graphics-core22` slot. Since `docker` and `nvidia-core22` share the same publisher, the interface gets auto connected if they are both installed. The `nvidia-core22` snap will not be automatically installed, since the plug doesn't set that as the default provider:
> 
> https://github.com/canonical/docker-snap/blob/2db03b40a188a03d0b38251875176cc16b05958b/snap/snapcraft.yaml#L50-L54
> 
> Docker snap has no other assertions for auto connecting its graphics plugs.
> 
> You can check the interface connections with: `snap connections docker`

@farshidtz , thank you for the response! That is very odd. We do not have nvidia-core22 installed, and this occurs even on fresh Ubuntu Core 24 images. We do have core22 and mesa-core22 installed, though (for some other snaps).

farshidtz commented:
> @farshidtz , thank you for the response! That is very odd. We do not have nvidia-core22 installed, and this occurs even on fresh Ubuntu Core 24 images. We do have core22 and mesa-core22 installed, though (for some other snaps).

Having mesa-core22 would have the same effect. It's published by the same publisher and the slot name matches: https://github.com/canonical/mesa-core22/blob/1a0ff3c93865ab7eb5bf8b754837c24fdd2933c7/snap/snapcraft.yaml#L328

Could you add a deny auto-connection rule from your dedicated store? If not, we'll have to slightly change the implementation of core24 Docker snap.

farshidtz committed:
Switch to stable pc-kernel
[skip ci]

farshidtz commented on snap/hooks/post-refresh:
Note sure why I can't comment on the lines. 
Line 21's second condition is the same as its first. Second one must be for core.

farshidtz commented on nvidia/system-detection:
```diff
@@ -0,0 +1,17 @@
+# Just exit if NVIDIA support is disabled #
+[ "${NVIDIA_SUPPORT_DISABLED}" != "true" ] || exit_inf "NVIDIA support disabled by user"
+
+NVIDIA_ENABLE="false"
+
+# Determine if we can enable NVIDIA support or not #
+[ "${NVIDIA_SUPPORT_CORE}" == "true" ] && NVIDIA_ENABLE="true"
+[ "${NVIDIA_SUPPORT_CLASSIC}" == "true" ] && NVIDIA_ENABLE="true"
+
+[ "${NVIDIA_ENABLE}" == "true" ] || exit_inf "No NVIDIA support detected"
```
Refactor: simplify using an if statement.

farshidtz commented on nvidia/nvidia-container-toolkit:
```diff
@@ -4,37 +4,27 @@ set -eu
-# Just exit if NVIDIA support is disabled #
-[ "${NVIDIA_SUPPORT_DISABLED}" != "true" ] || exit 0
+[ "${NVIDIA_SUPPORT_CLASSIC}" == "true" ] && echo "Running on Classic system"
+[ "${NVIDIA_SUPPORT_CORE}" == "true" ] && echo "Running on Ubuntu Core system"
 
-NVIDIA_ENABLE="false"
+# FIXME: error here if both core and classic are detecgted ?
+# [ "${NVIDIA_SUPPORT_CLASSIC}" == "true" -a "${NVIDIA_SUPPORT_CORE}" == "true"  ]
```
Refact: Log classic vs core, else error. 
Shouldn't this be done in system-detection script?

farshidtz commented on nvidia/lib:
```diff
@@ -78,7 +88,7 @@ nvidia_runtime_config () {
     # Default - opinionated, but most viable option for now #
     else
-        # FIXME: CDI spec-dirs can be set is a list using `"${SNAP_DATA}/etc/cdi",/var/run/cdi`, once this is fixed: https://github.com/NVIDIA/nvidia-container-toolkit/issues/466
+        # FIXME: CDI spec-dirs can be set is a list using `"${SNAP_DATA}/etc/cdi":/var/run/cdi`, once this is fixed: https://github.com/NVIDIA/nvidia-container-toolkit/issues/466
```
Check: this issue has been fixed.

farshidtz replied:
Comment dropped via https://github.com/canonical/docker-snap/pull/289

farshidtz commented on nvidia/lib:
```diff
@@ -58,9 +60,17 @@ cdi_generate () {
+    # UC22 #
+    [ -d "${SNAP}/graphics/lib/${ARCH_TRIPLET}" ] && CONTENT_PATH="${SNAP}/graphics"
+
+    # FIXME: if UC24, can we just rely on the wrapper env instead of these vars ?
```
Do envs from wrapper env match CDI_ ones?

farshidtz commented on nvidia/lib:
```diff
@@ -9,11 +9,13 @@ ARCH_TRIPLET="${U_MACHINE}-${U_KERNEL,,}-${U_USERLAND,,}"
+NVIDIA_SUPPORT_CLASSIC="false"
+[ -L "/var/lib/snapd/hostfs/usr/lib/${ARCH_TRIPLET}/libcuda.so" ] && NVIDIA_SUPPORT_CLASSIC="true"
+
+NVIDIA_SUPPORT_CORE="false"
+snapctl is-connected gpu-2404 && NVIDIA_SUPPORT_CORE="true"
+snapctl is-connected graphics-core22 && NVIDIA_SUPPORT_CORE="true"
```
Refact: conclude once first one passes.

farshidtz commented on nvidia/nvidia-container-toolkit-connect-hook:
```diff
@@ -0,0 +1,30 @@
+[ "${NVIDIA_SUPPORT_DISABLED}" != "true" ] || exit 0
+
+# Ensure nvidia support setup correctly #
+if snapctl is-connected "${PLUG_NAME}" ; then
```
Why check a connection in its connect hook? Isn't it always true?

farshidtz commented on snap/snapcraft.yaml:
```diff
@@ -120,6 +131,9 @@ parts:
     organize:
       nvidia/lib: usr/share/nvidia-container-toolkit/lib
       nvidia/nvidia-container-toolkit: bin/
+      nvidia/nvidia-container-toolkit-connect-hook: bin/
```
nvidia-container-toolkit-connect-hook isn't a "wrapper-script" (part's name).

farshidtz committed:
Fix graphics content path setting on core24 base snap (#274)

* Fix mesa-2204 content path override

* Prioritize use of gpu-2404

locnnil committed:
refact: change part name to match it's porpuse (#290)

Signed-off-by: Lincoln Wallace <lincoln.wallace@canonical.com>

locnnil committed:
fix: nvidia support check for Core in post-refresh (#291)

Signed-off-by: Lincoln Wallace <lincoln.wallace@canonical.com>

locnnil committed:
refact: simplify graphics provider checks (#288)

Signed-off-by: Lincoln Wallace <lincoln.wallace@canonical.com>

locnnil committed:
Merge branch 'main' into core24

locnnil committed:
Merge branch 'main' into core24

farshidtz commented on nvidia/lib:
```diff
@@ -9,12 +9,16 @@ ARCH_TRIPLET="${U_MACHINE}-${U_KERNEL,,}-${U_USERLAND,,}"
-if [ -L "/var/lib/snapd/hostfs/usr/lib/${ARCH_TRIPLET}/libcuda.so" ] ; then
-    NVIDIA_SUPPORT_CLASSIC="true"
+NVIDIA_SUPPORT_CLASSIC="false"
+[ -L "/var/lib/snapd/hostfs/usr/lib/${ARCH_TRIPLET}/libcuda.so" ] && NVIDIA_SUPPORT_CLASSIC="true"
+
+if snapctl is-connected gpu-2404 || snapctl is-connected graphics-core22; then
+    NVIDIA_SUPPORT_CORE="true"
 else
-    NVIDIA_SUPPORT_CLASSIC="false"
+    NVIDIA_SUPPORT_CORE="false"
 fi
```
With this, both NVIDIA_SUPPORT_CLASSIC and NVIDIA_SUPPORT_CORE can be true, when on classic with a provider installed and (auto-)connected. This is correctly handled in `nvidia/gpu-2404-optional-wrapper` because of the `exec` exit points but not in `nvidia/nvidia-container-toolkit`:


```console
$ snap connections docker
Interface          Plug                     Slot                                 Notes
...
content[gpu-2404]  docker:gpu-2404          mesa-2404:gpu-2404                   -
```


```console
$ snap logs docker.nvidia-container-toolkit -n 100
2025-06-17T18:19:32+02:00 docker.nvidia-container-toolkit[109211]: Running on Classic system
2025-06-17T18:19:32+02:00 docker.nvidia-container-toolkit[109211]: Running on Ubuntu Core system
```

locnnil committed:
doc: Drop comment regarding CDI spec-dirs setting (#289)

* refact: drop # FIXME comment

---------

Signed-off-by: Lincoln Wallace <lincoln.wallace@canonical.com>

locnnil committed:
refact: improve command chain scripts for nvidia (#293)

* refact(nvidia/gpu-2404-optional-wrapper): merge 'gpu-2024-optional-wrapper' into system-detection

* feat(nvidia/lib):
    - add shebang
    - add functions for detecting support on classic, core and in case of disabled support for nvidia 
    - Simplify detection of classic support using `nvidia_support_classic` function

* refact(nvidia/nvidia-container-toolkit): 
    - Move message of nvidia support on classic/core to nvidia/system-detection script

* refact(nvidia/nvidia-container-toolkit-connect-hook):
    - Use `nvidia_support_disabled` function to check if support for nvidia is active instead of env var

* feat(nvidia/system-detection):
    - exit on unbounded variables
    - Use `nvidia_support_classic`, `nvidia_support_core` and `nvidia_support_disabled` functions instead of env variables

* refact(snap/hooks/configure):
    - Use `nvidia_support_disabled` function instead of env variables

---------

Signed-off-by: Lincoln Wallace <lincoln.wallace@canonical.com>
Co-authored-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

locnnil commented:
I think @farshidtz has more information about this, but **as far I know and if I understand correctly** seems that the terms of CLA got upgraded/changed, and it's necessary to do a renewal/sign again. You can confirm this with the other CANONICAL folks in your meeting.

jocado commented:
CLA re-signed and submitted.  Just trying waiting to see if the processing on the Canonical side can be sped up at all.

farshidtz commented:
Thanks Justin for going through the new process. I've just triggered the check and it still fails. I'll follow up internally and let you know. This PR can wait a bit longer.

locnnil committed:
build: upgrade nvidia parts to v1.17.8 (#302)

Signed-off-by: Lincoln Wallace <lincoln.wallace@canonical.com>

farshidtz committed:
Add nvidia support instructions for Ubuntu Core 24 (#303)

* Add UC24 instructions

* Add note on ABI-compatibility

* Extend UC22 instructions

* Note on multiple providers

* Explain what is included in the snap

farshidtz committed:
Merge branch 'main' into core24

locnnil reviewed (approved):
I've taken another look, LGTM. +1
